fix parse_jobs_list_output parsing issues with SGE

QuantumChemist opened the pull request:
Addressing the issue https://github.com/Matgenix/qtoolkit/issues/50

- [x] remove parsing xml file generated by `qstat -xml -ext -j <JOB_ID>` option
- [ ] adjust job options for PBS and SGE in `_get_job_cmd`
- [x] move `memory_per_thread` to PBS base

QuantumChemist committed:
fix parse_jobs_list_output parsing issues

QuantumChemist committed:
removing some debugging prints

gpetretto commented:
Hi @QuantumChemist,

thanks a lot for looking into this. The main issue that I see is that having the current state of the Job is critical, at least for jobflow-remote (although I suppose it would be important for any other usage of qtoolkit). Since it is absolutely not clear how to go from this `JAT_status` to the standard states, to avoid spending too much time on this maybe we should give up for the moment on this and allow only the job list based on user for SGE. What do you think?

QuantumChemist commented:
> Hi @QuantumChemist,
> 
> thanks a lot for looking into this. The main issue that I see is that having the current state of the Job is critical, at least for jobflow-remote (although I suppose it would be important for any other usage of qtoolkit). Since it is absolutely not clear how to go from this `JAT_status` to the standard states, to avoid spending too much time on this maybe we should give up for the moment on this and allow only the job list based on user for SGE. What do you think?

Hi @gpetretto ,
actually simply parsing the jobs via the -u option sounds like the best approach for now.

I could change this PR then to remove the -j option in the new year (I'm on vacation already) or someone of you could go ahead if you don't want to wait. What would you prefer? 😃

gpetretto commented:
Hi @QuantumChemist,
sorry for the delay. I considered making this change, but I did not have the time. Do you think you can do this when you are back?

In addition, while doing the changes in #53 I notice some potential issues that I missed during the review of the previous PR.
1) The `_get_job_cmd` function:
https://github.com/Matgenix/qtoolkit/blob/62b683f345d8652a672cfec91949d353ad7a410d/src/qtoolkit/io/pbs_base.py#L119-L120
uses the `-j` option as well, but without the `-xml`. Is this working correctly? Should this also be removed for SGE now?
2) SGE adds the support to `memory_per_thread` and `gpus_per_job` for the qresources:
    https://github.com/Matgenix/qtoolkit/blob/62b683f345d8652a672cfec91949d353ad7a410d/src/qtoolkit/io/sge.py#L365
However, could `memory_per_thread` go in the base for pbs as well? And while `gpus_per_job` is marked as supported, it is not used in `_convert_qresources`, so it seems that it is basically not supported. Am I missing something?
3) In the generation of the test cases for `qdel`, is it correct that the `return_code` for the fifth and sixth case is 0? Also the error message is `stderr = b"qdel: job 269 deleted\nqdel: job 269 already completed\n"`, that contains both the successful output message and the error message. From the tests in the SGE container it seems that only `qdel: job 269 already completed` should be present. Is it different for the SGE version that you are using? (Same for `b"qdel: job 2675 deleted\nqdel: Invalid job id specified\n"`)

QuantumChemist committed:
undo previous commits

QuantumChemist committed:
only allow xml files creaed by the -u option

QuantumChemist committed:
implement the correct qstat call for SGE and PBS in _get_job_cmd

QuantumChemist committed:
moved memory_per_thread to PBSbase and deleted gpu_per_thread in SGE

QuantumChemist committed:
remove gpu option from unit tests

QuantumChemist commented:
Dear @gpetretto !
sorry for the late reply! I should have maybe mentioned that I took almost a month off 😅 .
I have now implemented almost all of the changes you mentioned. Thank you so much for catching those inconsistencies!
- regarding the `_get_job_cmd` function, I had accidentally only implemented the call in pbs_base.py that only works with SGE and doesn't for PBS, so I added an own  `_get_job_cmd` function for both (but I will ask a question to that in another thread)
- I have moved `memory_per_thread` to supported in pbs_base.py and removed `gpus_per_job` from sge.py, I think that was a sloppy copy&paste from slurm.py mistake, so thank you for catching that.

QuantumChemist commented on src/qtoolkit/io/pbs.py:
```diff
@@ -120,6 +120,10 @@ def _get_qstat_base_command(self) -> list[str]:
     def _get_job_ids_flag(self, job_ids_str: str) -> str:
         return job_ids_str
 
+    def _get_job_cmd(self, job_id: str):
+        cmd = f"qstat -f {job_id}"
+        return cmd
```
for PBS, I have restored the previous state of this function, but I wonder if we want to switch to xml output like that
```suggestion
    def _get_job_cmd(self, job_id: str):
        cmd = f"qstat -fx {job_id}"
        return cmd
```
maybe?

gpetretto replied:
Thanks for restoring that. 
I am not sure if PBS has an xml mode. Based on the documentation and on an OpenPBS contianer that I have set up, the `-x` option just adds jobs that have been completed to the list, but always in the standard format. Even if it was possible to switch to an xml format, this would require additional work that I am not sure would be worth at the moment. 
I would suggest to leave PBS unchanged.

QuantumChemist replied:
Ok, then I will leave it like that!

QuantumChemist commented on src/qtoolkit/io/sge.py:
```diff
@@ -280,6 +280,10 @@ def _get_qstat_base_command(self) -> list[str]:
     def _get_job_ids_flag(self, job_ids_str: str) -> str:
         return f"-j {job_ids_str}"
 
+    def _get_job_cmd(self, job_id: str):
+        cmd = f"qstat -j {job_id}"
+        return cmd
```
In case of SGE, I'm not sure if I understood it correctly if we always need the `-u` option (instead of `-j`)?
Otherwise, also here we could parse this as an xml file maybe? 
```suggestion
    def _get_job_cmd(self, job_id: str):
        cmd = f"qstat -j {job_id} -xml"
        return cmd
```

QuantumChemist replied:
Otherwise, I will switch this to the -u option.

gpetretto replied:
I think that the `-xml` option can indeed be moved to  `_get_job_cmd` for SGE

QuantumChemist reviewed (commented):
these are my questions regarding the `_get_job_cmd` function

QuantumChemist commented:
and lastly regarding the unit tests:
I have written the files "manually" (with ChatGPT) because I don't have SGE on any cluster available and have used the slurm unit tests as a guide:
https://github.com/Matgenix/qtoolkit/blob/62b683f345d8652a672cfec91949d353ad7a410d/tests/test_data/io/slurm/create_parse_cancel_output_inout.py#L71
https://github.com/Matgenix/qtoolkit/blob/62b683f345d8652a672cfec91949d353ad7a410d/tests/test_data/io/slurm/create_parse_cancel_output_inout.py#L86
So I have basically just copied this behaviour 😅

gpetretto commented on src/qtoolkit/io/sge.py:
```diff
@@ -294,7 +298,21 @@ def parse_jobs_list_output(self, exit_code, stdout, stderr) -> list[QJob]:
+            )
+
         job_elements = xmldata.getElementsByTagName("job_list")
+        if not job_elements:
```
I am afraid this check will fail in case there are no jobs in the queue. In the SGE container an empty list of jobs looks like this

```
<?xml version='1.0'?>
<job_info  xmlns:xsd="http://arc.liv.ac.uk/repos/darcs/sge/source/dist/util/resources/schemas/qstat/qstat.xsd">
  <queue_info>
  </queue_info>
  <job_info>
  </job_info>
</job_info>
```
I think it would be safer to remove it

QuantumChemist replied:
I have removed this check. I guess checking for the `job_info` tag should be sufficient as this tag doesn't exist when the job info is queried using the job_id(s) and the -j option (a `djob_info` tag exists instead).

gpetretto reviewed (commented):
Hi @QuantumChemist,

no worries about the delay. I hope you enjoyed your holidays!

Aside from the comments to the code, here are more general comments:
* I would remove entirely the  possibility to query by job ids for SGE. So, moving `_get_jobs_list_cmd` to the subclasses and for SGE raise if a list of values is passed to `job_ids`.
* For the tests, just relying on the manually generated unit tests seems to be a strong limitation. The best will be to have integration tests similarly to jobflow-remote and unit tests generated from actual execution on the containers. So I propose to leave the SGE tests as they are for the moment and we will review the whole testing procedure for all the schedulers at a later stage.

Is this fine with you?

QuantumChemist committed:
add xml option to _get_job_cmd in sge.py

QuantumChemist committed:
remove OutputParsingError for job_elements in parse_jobs_list_output in sge.py

QuantumChemist committed:
adjust _get_job_list_cmd

QuantumChemist committed:
adjust unit tests

QuantumChemist committed:
fix typo in comment

QuantumChemist commented on src/qtoolkit/io/sge.py:
```diff
@@ -280,6 +293,10 @@ def _get_qstat_base_command(self) -> list[str]:
     def _get_job_ids_flag(self, job_ids_str: str) -> str:
         return f"-j {job_ids_str}"
 
+    def _get_job_cmd(self, job_id: str):
+        cmd = f"qstat -j -xml {job_id}"
```
the last thing that I am not sure if I understand it correctly is if I shall change this to:
```suggestion
    def _get_job_cmd(self, user: str):
        cmd = f"qstat -u -xml {user}"
```

to make it entirely impossible to query by (a list of) job ID(s)

gpetretto replied:
Thanks for the updates. Concerning this point, this is already defined in the original base class (`BaseSchedulerIO`) and the input is a job_id. So not to break the inheritance here I would say that it should stay that way. I propose to also raise here, saying that this option is not supported for SGE

QuantumChemist reviewed (commented):
Hi @gpetretto !

Thank you for the feedback! I have made the changes you ask for, except for a last question about `_get_job_cmd` (see below).

And left the unit test data as it currently is then. 😃

QuantumChemist committed:
remove querying by job IDs support for SGE

QuantumChemist committed:
adjust unit tests

QuantumChemist commented on src/qtoolkit/io/sge.py:
```diff
@@ -278,7 +291,10 @@ def _get_qstat_base_command(self) -> list[str]:
         return ["qstat", "-ext", "-urg", "-xml"]
 
     def _get_job_ids_flag(self, job_ids_str: str) -> str:
-        return f"-j {job_ids_str}"
+        raise NotImplementedError("Querying by job IDs is not supported for SGE.")
+
+    def _get_job_cmd(self, job_id: str):
+        raise NotImplementedError("Querying by job IDs is not supported for SGE.")
```
I have added the error raise concerning the job_IDs at this two places in the code now.

QuantumChemist reviewed (commented):
Hi @gpetretto ,

let me know if more changes are needed!

gpetretto reviewed (approved):
Thanks @QuantumChemist!
